internal/notifier/datadog.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2023 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package notifier
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadog"
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadogV1"
+
+	eventv1 "github.com/fluxcd/pkg/apis/event/v1beta1"
+)
+
+type DataDog struct {
+	apiClient *datadog.APIClient
+	eventsApi *datadogV1.EventsApi
+	apiKey    string
+	env       string
+}
+
+// NewDataDog creates a new DataDog provider by mapping the notification provider API to sensible values for the DataDog API.
+// url: The DataDog API endpoint to use. Examples: https://api.datadoghq.com, https://api.datadoghq.eu, etc.
+// token: The DataDog API key (not the application key).
+// channel: The name of the Kubernetes cluster that is producing these events, mapped to the `env` tag in DataDog.
+// headers: A map of extra tags to add to the event
+func NewDataDog(address string, proxyUrl string, certPool *x509.CertPool, token string, channel string) (*DataDog, error) {
+	conf := datadog.NewConfiguration()
+
+	if token == "" {
+		return nil, fmt.Errorf("token cannot be empty")
+	}
+
+	if channel == "" {
```

Comment:
Should the channel be required? It seems that the Datadog API doesn't require tags

Reply:
The `env` tag is usually used all over datadog (for metrics and logs) in order to enable [unified service tagging](https://docs.datadoghq.com/getting_started/tagging/unified_service_tagging), but you're right that it isn't a strict requirement on DataDog's side.

I understand your point about not necessarily needing to make this a requirement.  I made this decision before I realized that the event metadata via the `Alert` can be used by the user to add any tags they want.  So I'm okay with removing this requirement and expecting users to use event metadata to add this tag.

Reply:
I think we can still have the `channel` set to `env` tag if it is specied. 
It just wouldn't be required for those who don't care about the tag.

Reply:
I thought about it some more, and I actually think it's better to just not use `channel` at all, and expect users to set tags like `env` via the event metadata.  Is that approach alright?